Snowflake.Data/Core/CredentialManager/Infrastructure/SFCredentialManagerFileImpl.cs

```diff
@@ -0,0 +1,157 @@
+
+        internal SFCredentialManagerFileImpl(FileOperations fileOperations, DirectoryOperations directoryOperations, UnixOperations unixOperations, EnvironmentOperations environmentOperations)
+        {
+            _fileOperations = fileOperations;
+            _directoryOperations = directoryOperations;
+            _unixOperations = unixOperations;
+            _environmentOperations = environmentOperations;
+            SetCredentialCachePath(ref _jsonCacheDirectory, ref _jsonCacheFilePath);
+        }
+
+        private void SetCredentialCachePath(ref string _jsonCacheDirectory, ref string _jsonCacheFilePath)
+        {
+            var customDirectory = _environmentOperations.GetEnvironmentVariable(CredentialCacheDirectoryEnvironmentName);
+            _jsonCacheDirectory = string.IsNullOrEmpty(customDirectory) ? Path.Combine(HomeDirectoryProvider.HomeDirectory(_environmentOperations), CredentialCacheDirName) : customDirectory;
+            if (!_directoryOperations.Exists(_jsonCacheDirectory))
+            {
+                _directoryOperations.CreateDirectory(_jsonCacheDirectory);
+            }
+            _jsonCacheFilePath = Path.Combine(_jsonCacheDirectory, CredentialCacheFileName);
+            s_logger.Info($"Setting the json credential cache path to {_jsonCacheFilePath}");
+        }
+
+        internal void WriteToJsonFile(string content)
+        {
+            s_logger.Debug($"Writing credentials to json file in {_jsonCacheFilePath}");
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                _fileOperations.Write(_jsonCacheFilePath, content);
+                FileInfo info = new FileInfo(_jsonCacheFilePath);
+                FileSecurity security = info.GetAccessControl();
+                FileSystemAccessRule rule = new FileSystemAccessRule(
+                    new SecurityIdentifier(WellKnownSidType.CreatorOwnerSid, null),
+                    FileSystemRights.FullControl,
+                    AccessControlType.Allow);
+                security.SetAccessRule(rule);
+                info.SetAccessControl(security);
+            }
+            else
+            {
+                if (!_directoryOperations.Exists(_jsonCacheDirectory))
+                {
+                    _directoryOperations.CreateDirectory(_jsonCacheDirectory);
+                }
+                s_logger.Info($"Creating the json file for credential cache in {_jsonCacheFilePath}");
+                if (_fileOperations.Exists(_jsonCacheFilePath))
+                {
+                    s_logger.Info($"The existing json file for credential cache in {_jsonCacheFilePath} will be overwritten");
+                }
+                var createFileResult = _unixOperations.CreateFileWithPermissions(_jsonCacheFilePath,
```

Comment:
is overriding of existing file also implemented in other drivers?
I'm thinking of using a file handle like that for file operations and permission checks like in that example:

```
        var fileInfo = new UnixFileInfo(path: path);

        using var handle = fileInfo.OpenRead();
        if (handle.OwnerUser.UserId != Syscall.geteuid())
            throw new SecurityException("Attempting to read a file not owned by the effective user of the current process");
        if (handle.OwnerGroup.GroupId != Syscall.getegid())
            throw new SecurityException("Attempting to read a file not owned by the effective group of the current process");
        if ((handle.FileAccessPermissions & forbiddenPermissions) != 0)
            throw new SecurityException("Attempting to read a file with too broad permissions assigned");

        using var streamReader = new StreamReader(handle, Encoding.Default);
        return streamReader.ReadToEnd();
```

Reply:
The other connector also overwrites the file when saving a credential

Reply:
Also added the security checks from your example when reading the file